Create Push apps and variants in UPS based on equivalent CRs

grdryn opened the pull request:
JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

----

## Verification steps

1. Deploy the operator:

```
make cluster/prepare
kubectl apply -n unifiedpush -f ./deploy/operator.yaml
```

2. Provision a UPS by creating a CR:

```
kubectl apply -n unifiedpush -f ./deploy/crds/push_v1alpha1_unifiedpushserver_cr.yaml
```

3. Create a PushApplication CR:

```
kubectl apply -n unifiedpush -f ./deploy/crds/push_v1alpha1_pushapplication_cr.yaml
```

- Ensure the PushApplication shows up in the list of applications in the UPS admin UI
- Ensure the PushApplication has the `pushApplicationId` and `masterSecret` fields populated in the CR status.

4. Create an AndroidVariant CR using the PushApplication ID:

```
cat deploy/crds/push_v1alpha1_androidvariant_cr.yaml | sed -e "s/pushApplicationId: '123456'/pushApplicationId: '$(kubectl get pushapplications -o jsonpath='{.items[0].status.pushApplicationId}')'/" | kubectl apply -f -
```

- Ensure the variant shows up under the Application in the UPS admin UI
- Ensure the AndroidVariant CR has a `status.variantId` populated, and `status.ready` is `true`
- Ensure the AndroidVariant CR has a `finalizer.push.aerogear.org` in `metadata.finalizers`

*Note:* You can do something equivalent to this for the IOSVariant if you've got a cert that you can use

5. Delete the AndroidVariant CR:

```
kubectl delete androidVariant example-androidvariant
```

- Ensure the Variant no longer appears in the UPS admin UI
- Ensure the CR is no longer accessible in Kubernetes

*Note:* You can do something equivalent to this for the IOSVariant if you've created one in step 4.

6. Delete the PushApplication CR:

```
kubectl delete pushApplication example-pushapplication
```

- Ensure the application no longer appears in the UPS admin UI
- Ensure the CR is no longer accessible in Kubernetes

7. Clean up!

```
make cluster/clean
```

aliok committed:
operator-sdk add controller --api-version=push.aerogear.org/v1alpha1 --kind=PushApplication

aliok committed:
operator-sdk add controller --api-version=push.aerogear.org/v1alpha1 --kind=AndroidVariant

aliok committed:
operator-sdk add controller --api-version=push.aerogear.org/v1alpha1 --kind=IOSVariant

aliok committed:
Fill push app controller

grdryn committed:
Add CreateApplication method to UPS client

grdryn committed:
Add new CRDs to "make cluster/prepare"

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Add missing internal Service to provisioned UPS

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

This is needed for the operator to be able to interact with UPS within
the cluster, without needing to go through the oauth-proxy.

This also adds a new label to it with a key of "internal" and a value
of "unifiedpush", this is for filtering the correct one more easily.

grdryn committed:
Get valid client in PushApplication controller

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

This means finding all (hopefully just one) Services that are labelled
with internal=unifiedpush, and creating a UnifiedPushClient with the
first one.

Since the Service won't be directly routable from a local instance of
the operator, instructions have also been added for a possible way to
make it available locally.

grdryn committed:
Add finalizer to PushApplication to allow us to delete from UPS

grdryn committed:
Delete PushApplication from UPS when CR deleted

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Add GET and DELETE methods for variants in client

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Create/Delete variants in UPS in response to CRs

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Add some docs around the CRs to the README.adoc

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Set status.ready to true in Variant CRs

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn committed:
Create and delete all CRDs properly in Makefile

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

grdryn commented on pkg/controller/util/add_finalizer.go:
```diff
@@ -0,0 +1,35 @@
+
+		// Update CR
+		switch typedObject := o.(type) {
+		case runtime.Object:
```
This entire type switch is just shenanigans because the `metav1.Object` and `runtime.Object` interfaces for some reason are not aligned nicely.

aliok replied:
a type assertion should be better here IMO: https://tour.golang.org/methods/15

aliok replied:
```
        runtimeObject, ok := o.(runtime.Object)
	if !ok{
		reqLogger.Info("Can't determine the type of thing to add finalizer to")
		return nil
	}
```

aliok replied:
or, the entire method:

```golang
// AddFinalizer will add a finalizer to the PushApplication CR so that
// we can delete from UPS appropriately
func AddFinalizer(client client.Client, reqLogger logr.Logger, o metav1.Object) error {
	// This is based on the example code at:
	// https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#handle-cleanup-on-deletion

	if len(o.GetFinalizers()) > 0 || o.GetDeletionTimestamp() != nil {
		return nil
	}

	reqLogger.Info("Adding Finalizer to the PushApplication")
	o.SetFinalizers([]string{"finalizer.push.aerogear.org"})

	runtimeObject, ok := o.(runtime.Object)
	if !ok {
		reqLogger.Info("Can't determine the type of thing to add finalizer to")
		return nil
	}

	err := client.Update(context.TODO(), runtimeObject)
	if err != nil {
		reqLogger.Error(err, "Failed to update a CR with a finalizer")
		return err
	}

	return nil
}
```

grdryn replied:
Ah yes, that's nicer, I'll update it now, thanks!

grdryn commented on pkg/controller/util/unifiedpushClient.go:
```diff
@@ -0,0 +1,35 @@
+		return unifiedpush.UnifiedpushClient{}, errors.New("Didn't find any UPS Services")
+	}
+
+	if len(foundServices.Items) > 1 {
```
I've created [AEROGEAR-9335](https://issues.jboss.org/browse/AEROGEAR-9335) to only support provisioning of a single UPS instance (when shouldn't provision a second one even if a second CR is created).

grdryn commented on pkg/unifiedpush/client.go:
```diff
@@ -124,6 +195,53 @@ func (c UnifiedpushClient) CreateAndroidVariant(av *pushv1alpha1.AndroidVariant)
+func (c UnifiedpushClient) DeleteAndroidVariant(v *pushv1alpha1.AndroidVariant) error {
+	if v.Status.VariantId == "" {
+		// We haven't created it yet
+		return nil
```
Not sure if this should return an error. I'm not convinced that an error would be useful.

grdryn commented:
I have built this locally and pushed to `quay.io/aerogear/unifiedpush-operator:PR-23`.

aliok commented:
@grdryn what's remaining here?

grdryn commented:
@aliok I think this covers everything for AEROGEAR-9022 (and AEROGEAR-9300) now.

Maybe something should be added to the e2e test, but I'm not sure how important that is, since that doesn't currently run on PRs (cc @psturc).

We're also really low on testing in general, like there are no unit tests. I think it might be better to add improvements like that asynchronously in separate issues/PRs, to allow the functionality here to unblock other tasks in the story. WDYT?

aliok commented:
@grdryn 

> @aliok I think this covers everything for AEROGEAR-9022 (and AEROGEAR-9300) now.

Ok, will give it a final review and a verification and then let's merge!

> I think it might be better to add improvements like that asynchronously in separate issues/PRs, to allow the functionality here to unblock other tasks in the story. WDYT?

+1

grdryn commented on Makefile:
```diff
@@ -44,12 +44,18 @@ cluster/prepare:
 
 .PHONY: cluster/clean
 cluster/clean:
 	-kubectl delete -f deploy/role.yaml
```
Because there are now finalizers on the PushApplication, AndroidVariant, and IOSVariant CRs, those will have to be cleaned up here first before anything else (otherwise manual intervention will be needed to delete them to allow the namespace to be deleted).

Thanks to @psturc for highlighting that. I'll add it now.

grdryn replied:
Done in latest push.

psturc replied:
👍 works perfectly

grdryn committed:
Delete all CRs first in cluster/clean

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

This is so that the finalizers will get dealt with properly, and the
namespace can eventually be deleted successfully.

aliok commented:
@grdryn I added a comment to change something.

other than that, everything looks good.
I also created/deleted bunch of push apps and also variants and it all looked fine.
let's merge this! ...once that simple change I requested is done

psturc commented on README.adoc:
```diff
@@ -83,6 +86,165 @@ kubectl label namespace unifiedpush monitoring-key=middleware
+kubectl get PushApplication example-pushapplication -n unifiedpush -o jsonpath='{.status.pushApplicationId}'
+....
+
+Here are all of the configurable fields in an AndroidVariant:
```
```suggestion
Here are all of the configurable fields in an IOSVariant:
```

grdryn replied:
Fixed in 48d616c, thanks!

psturc reviewed (approved):
Just a small typo to fix, otherwise LGTM!
We can extend e2e tests in a separate task.

grdryn committed:
Use type assertion rather than type switch

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

Co-authored-by: Ali Ok <aliok@redhat.com>

grdryn committed:
Fix typo in README: AndroidVariant -> IOSVariant

JIRA: https://issues.jboss.org/browse/AEROGEAR-9022

Co-authored-by: Pavel Šturc <psturc@redhat.com>

grdryn commented:
Ooh, look what happened when I added you guys as "Co-authored-by" in the commits based to address your comments/suggestions! :smiley: 

![image](https://user-images.githubusercontent.com/442386/59359950-38575800-8d27-11e9-95b1-cbb200a528da.png)

aliok reviewed (approved):
MERGE IT!

grdryn commented:
Sorry, I meant to merge it yesterday! :see_no_evil:
